aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java

```diff
@@ -147,25 +149,56 @@ public Completable start() {
                     subscriptionProcessor.startDrainingMutationBuffer();
                 }
                 status.compareAndSet(OrchestratorStatus.STARTING, OrchestratorStatus.STARTED);
+                LOG.debug("Orchestrator started.");
             })
-        );
-        return initializationCompletable;
+        ).doFinally(startStopSemaphore::release);
     }
 
     /**
      * Stop all model synchronization.
      */
-    public void stop() {
-        if (isStarted()) {
+    public synchronized void stop() {
+        if (!performAction(OrchestratorAction.STOP)) {
+            return;
+        }
+        try {
             LOG.info("Intentionally stopping cloud synchronization, now.");
-            status.compareAndSet(OrchestratorStatus.STARTED, OrchestratorStatus.STOPPING);
             subscriptionProcessor.stopAllSubscriptionActivity();
             storageObserver.stopObservingStorageChanges();
             mutationProcessor.stopDrainingMutationOutbox();
             status.compareAndSet(OrchestratorStatus.STOPPING, OrchestratorStatus.STOPPED);
             LOG.debug("Stopped remote synchronization.");
+        } finally {
+            startStopSemaphore.release();
         }
+    }
 
+    private boolean performAction(OrchestratorAction intendedAction) {
+        OrchestratorStatus expectedCurrentStatus = OrchestratorAction.START.equals(intendedAction) ?
+            OrchestratorStatus.STOPPED : OrchestratorStatus.STARTED;
+        try {
+            LOG.debug(String.format("Trying to %s the orchestrator.", intendedAction.name()));
+            boolean permitAcquired = startStopSemaphore.tryAcquire(ACQUIRE_PERMIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            if (!permitAcquired) {
+                LOG.warn(String.format("Unable to acquire permit to %s the orchestrator. ", intendedAction.name()));
+                return false;
+            }
+            boolean statusSet = status.compareAndSet(intendedAction.expectedCurrentStatus, intendedAction.targetStatus);
+            // only stop if it's started AND if we can get a permit.
+            if (!statusSet) {
+                LOG.warn(String.format("Failed %s orchestrator. Current status: %s",
```

Comment:
This might be an error, if the user is expecting cloud sync to function? Should we wire this back to the user error handler somehow, too?

Reply:
I added some logic here to invoke the user-provided onError callback (if one is provided)